src/rpc/client/blobs.rs

```diff
@@ -0,0 +1,1795 @@
+//! - [`read`](Client::read) creates a [reader](Reader) to read data from.
+//! - [`export`](Client::export) eports data to a file on the local filesystem.
+//!
+//! ## Interacting with remote nodes
+//!
+//! - [`download`](Client::download) downloads data from a remote node.
+//!   remote node.
+//!
+//! ## Interacting with the blob store itself
+//!
+//! These are more advanced operations that are usually not needed in normal
+//! operation.
+//!
+//! - [`consistency_check`](Client::consistency_check) checks the internal
+//!   consistency of the local blob store.
+//! - [`validate`](Client::validate) validates the locally stored data against
+//!   their BLAKE3 hashes.
+//! - [`delete_blob`](Client::delete_blob) deletes a blob from the local store.
+//!
+//! ### Batch operations
+//!
+//! For complex update operations, there is a [`batch`](Client::batch) API that
+//! allows you to add multiple blobs in a single logical batch.
+//!
+//! Operations in a batch return [temporary tags](crate::util::TempTag) that
+//! protect the added data from garbage collection as long as the batch is
+//! alive.
+//!
+//! To store the data permanently, a temp tag needs to be upgraded to a
+//! permanent tag using [`persist`](crate::rpc::client::blobs::Batch::persist) or
+//! [`persist_to`](crate::rpc::client::blobs::Batch::persist_to).
+use std::{
+    future::Future,
+    io,
+    path::PathBuf,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use anyhow::{anyhow, Context as _, Result};
+use bytes::Bytes;
+use futures_lite::{Stream, StreamExt};
+use futures_util::SinkExt;
+use genawaiter::sync::{Co, Gen};
+use iroh_net::NodeAddr;
+use portable_atomic::{AtomicU64, Ordering};
+use quic_rpc::{
+    client::{BoxStreamSync, BoxedConnector},
+    Connector, RpcClient,
+};
+use serde::{Deserialize, Serialize};
+use tokio::io::{AsyncRead, AsyncReadExt, ReadBuf};
+use tokio_util::io::{ReaderStream, StreamReader};
+use tracing::warn;
+
+pub use crate::net_protocol::DownloadMode;
+use crate::{
+    export::ExportProgress as BytesExportProgress,
+    format::collection::{Collection, SimpleStore},
+    get::db::DownloadProgress as BytesDownloadProgress,
+    net_protocol::BlobDownloadRequest,
+    rpc::proto::RpcService,
+    store::{BaoBlobSize, ConsistencyCheckProgress, ExportFormat, ExportMode, ValidateProgress},
+    util::SetTagOption,
+    BlobFormat, Hash, Tag,
+};
+
+mod batch;
+pub use batch::{AddDirOpts, AddFileOpts, AddReaderOpts, Batch};
+
+use super::{flatten, tags};
+use crate::rpc::proto::blobs::{
+    AddPathRequest, AddStreamRequest, AddStreamUpdate, BatchCreateRequest, BatchCreateResponse,
+    BlobStatusRequest, ConsistencyCheckRequest, CreateCollectionRequest, CreateCollectionResponse,
+    DeleteRequest, ExportRequest, ListIncompleteRequest, ListRequest, ReadAtRequest,
+    ReadAtResponse, ValidateRequest,
+};
+
+/// Iroh blobs client.
+#[derive(Debug, Clone)]
+pub struct Client<C = BoxedConnector<RpcService>> {
```

Comment:
why is this not repr transparent?

Reply:
I don't need it in this direction. But you never know... So going to add it.